File: src-p/infra/adapters/proxy/ReplaceEngine.py
import re
from typing import List
from domain.proxy.IReplaceEngine import IReplaceEngine
from domain.proxy.ProxyRules import ReplaceRule


class ReplaceEngine(IReplaceEngine):
    def apply(self, text: str, rules: List[ReplaceRule], phase: str) -> str:
        for rule in rules:
            target = rule.on or "response"
            if target != phase and target != "both":
                continue

            if rule.from_.startswith("regex:"):
                pattern = rule.from_[6:]
                text = re.sub(pattern, rule.to, text, flags=re.IGNORECASE)
            else:
                text = text.replace(rule.from_, rule.to)

        return text

File: src/infra/adapters/proxy/reverse/ReplaceEngine.ts
import { IReplaceEngine } from "../../../../domain/proxy/reverse/IReplaceEngine.js";
import { ReplaceRule } from "../../../../domain/proxy/reverse/ProxyRules.js";

export class ReplaceEngine implements IReplaceEngine {
  apply(text: string, rules: ReplaceRule[], phase: "request" | "response"): string {
    for (const rule of rules) {
      const target = rule.on ?? "response";
      if (target !== phase && target !== "both") continue;

      if (rule.from.startsWith("regex:")) {
        text = text.replace(new RegExp(rule.from.slice(6), "gi"), rule.to);
      } else {
        text = text.replaceAll(rule.from, rule.to);
      }
    }
    return text;
  }
}

